File: basic/src/main/java/top/non/late/pavilion/basic/java/designs/builder/code/BookDirector.java
package top.non.late.pavilion.basic.java.designs.builder.code;

/**
 * Created by nonlatepavilion@outlook.com on 2017/9/7.
 * Desc: 指挥者。包含 builder，用来统一product的生产过程
 */
public class BookDirector {
    private BookBuilder builder;

    public BookDirector(BookBuilder builder) {
        this.builder = builder;
    }

    public BookProduct build() {
        builder.bookName();
        builder.author();
        builder.isbn();
        return builder.build();
    }

}

File: basic/src/main/java/top/non/late/pavilion/basic/kotlin/designs/builder/BookDirector.kt
package top.non.late.pavilion.basic.kotlin.designs.builder

/**
 * Created by nonlatepavilion@outlook.com on 2017/9/25.
 * Desc: 
 */
class BookDirector(private val builder: BookBuilder) {

    fun build(): BookProduct {
        builder.bookName()
        builder.author()
        builder.isbn()
        return builder.build()
    }
}
